### ISSUE TITLE
oss-fuzz does not cover the code after if - else in C code

### ISSUE THREAD

**Issue by ezik117** [google/oss-fuzz#11788]
I created a simple C project with the only function with 3 branches (if incoming values equal 0, 32 and whatever else). Here is the testing function:

```
1  int func(int a)
2  {
3     FILE *fptr;
4     if ((fptr = fopen("tdata.bin","ab")) != NULL)
5     {
6         fwrite(&a, 1, 1, fptr);
7         fclose(fptr);
8     }
9
10    if (a == 0x00)
11        return 0;
12    else if (a == 0x20)
13        return 1;
14
15    printf("last line is reached!");
16    return 2;
17 }
```
Here is the fuzzer:

```
int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
{
  func((int)(*data));
  return 0;
}
```
The issue that if I run the introspector test with the following command:

```
python3 infra/helper.py introspector test02
```

it shows me that the function coverage is only 90%. The lines 15-16 are not covered. Although the fuzzing is passing those lines. It can be seen in output console (I can see "last line is reached!") and also in the debug file "tdata.bin" I can see about 562k randomized bytes with values not equal 0 and 32. There are also two corpuses with values 0 and 32. So it seems the fuzzer is reached the lines 15-16 but for some reason does not count them in the report.

**Comment by DavidKorczynski** [comment]
Could be related to some optimization issue. Can you share the full OSS-Fuzz project source code you use? including Dockerfile and `build.sh`

**Comment by ezik117** [comment]
Hello David! Here is the repo: https://github.com/ezik117/test02
Also if you will comment out all lines in the func() except 'printf' and 'return', the Project function overview section will show '85.71%' for the 'func'.

**Comment by DavidKorczynski** [comment]
It's reported as 85% percent because the corpus created does not exercise all paths in the module.

The corpus is:

```sh
$ ls build/corpus/tester/test02/
5ba93c9db0cff93f52b521d7420e43f6eda2784f  b858cb282617fb0956d960215c8e84d1ccf909c6
```

The coverage is:

![Screenshot 2024-04-09 222824](https://github.com/google/oss-fuzz/assets/657617/c9fc83b3-7b94-4f59-a16e-348e4e0bbf34)

I.e. there are only two paths taken by the corpus accumulated.

I think it's likely some corpus pruning or similar that removes the seed that triggers the printf.

In short, Fuzz Introspector does what it should, but there is some other issue that happens which causes the seed for the third parth to not be in the corpus directory, and I think it's likely some pruning stuff.

**Comment by ezik117** [comment]
Yes, I think your suggestion about the seed is correct. But I just tried to convert the function into following:
```
void func(int a)
{
    void* m1;
    void* m2;
    switch (a) {
        case 0:
            m1 = malloc(a);
            // do something here, for example send data through socket
            free(m1);
            break;
        case 1:
            m2 = malloc(a + 1);
            // another action, let's say saving the file
            free(m2);
            break;
        
        default:
            break;
    }
}
```
The code without 'return' in the middle. And as expected the function was considered as blocked. Although it work just fine and with different seed goes deeper and deeper over switch statement. So, to consider the function as 'well developed' for code coverage analysis the outline must follow the standard "put return in the middle, normal exit will be omitted"? Correct me, please, I maybe be wrong in my suggestion...

**Comment by ezik117** [comment]
I know the code looks stupid, but I am automation testing engeneer, and my task is to ruin things that looks right :) This time I found that code coverage is not behave as expected, this is why I try to understand is my finding correct or maybe I am wrong?

**Comment by DavidKorczynski** [comment]
I wouldn't conclude that "code coverage is not behave as expected" -- I think many things can play a role here, in particular optimizations by the compiler on the code which forces the underlying bytecode to not be in the same way as the code looks, which ultimately causes the feedback mechanism of the fuzzing engine to not produce seeds as expected.

However, do you have real-world examples?

Ultimately, this doesn't seem to be an OSS-Fuzz issue but rather fuzzer engines / coverage extraction logic.

**Comment by DavidKorczynski** [comment]
Closing this one as it's not really an OSS-Fuzz issue as addressed above. Please feel free to re-open if you believe there is an error on OSS-Fuzz's side here.
